operators/src/concurrency/thread_pool.rs

```diff
@@ -0,0 +1,339 @@
+use std::sync::Arc;
+use std::thread::JoinHandle;
+use std::{iter, mem};
+
+use crossbeam::atomic::AtomicCell;
+use crossbeam::deque::{Injector, Steal, Stealer, Worker};
+use crossbeam::sync::WaitGroup;
+use crossbeam::utils::Backoff;
+use failure::_core::marker::PhantomData;
+
+/// A chunk of work with some metadata
+struct Task {
+    _group_id: TaskGroupId,
+    task_fn: Box<dyn TaskFn>,
+}
+
+type TaskGroupId = usize;
+
+pub trait TaskFn: FnOnce() + Send {}
+pub trait StaticTaskFn: TaskFn + 'static {}
+
+impl<T> TaskFn for T where T: FnOnce() + Send {}
+impl<T> StaticTaskFn for T where T: TaskFn + 'static {}
+
+impl Task {
+    /// Create a new task to be executed at some point
+    fn new<F>(group_id: TaskGroupId, f: F) -> Self
+    where
+        F: StaticTaskFn,
+    {
+        Self {
+            _group_id: group_id,
+            task_fn: Box::new(f),
+        }
+    }
+
+    /// Executes the task
+    /// TODO: use `FnTraits` once stable
+    fn call_once(self) {
+        (self.task_fn)()
+    }
+}
+
+/// A worker thread pool for compute-heavy tasks
+///
+/// TODO: increase threads to a certain maximum number if current threads don't produce results fast enough?
+/// TODO: schedule tasks based on group context?
+///
+#[derive(Debug)]
+pub struct ThreadPool {
+    degree_of_parallelism: usize,
+    global_queue: Arc<Injector<Task>>,
+    stealers: Vec<Stealer<Task>>,
+    threads: Vec<JoinHandle<()>>,
+    next_group_id: AtomicCell<usize>,
+}
+
+impl ThreadPool {
+    pub fn new(number_of_threads: usize) -> Self {
+        assert!(
+            number_of_threads > 0,
+            "There must be at least one thread for the thread pool"
+        );
+
+        let worker_deques: Vec<Worker<Task>> =
+            (0..number_of_threads).map(|_| Worker::new_fifo()).collect();
+
+        let mut thread_pool = Self {
+            degree_of_parallelism: number_of_threads,
+            global_queue: Arc::new(Injector::new()),
+            stealers: worker_deques.iter().map(Worker::stealer).collect(),
```

Comment:
please explain (or point me to resources) of why you have a global AND worker-local queues. Why not use a global queue where all workers get work from?

Reply:
Because all workers would poll from the same resource all the time, which means you have more lock conflicts. By doing this, you minimize the conflicts on one queue.